signer/src/transaction_coordinator.rs

```diff
@@ -2606,14 +2603,33 @@ pub async fn should_coordinate_dkg(
 
 /// Assert, given the last dkg and smart contract current aggregate key, if we
 /// need to verify the shares and/or issue a rotate key call.
-pub fn assert_rotate_key_action(
+pub async fn assert_rotate_key_action<C>(
+    context: &C,
     last_dkg: &model::EncryptedDkgShares,
     current_aggregate_key: Option<PublicKey>,
-) -> Result<(bool, bool), Error> {
+) -> Result<(bool, bool), Error>
+where
+    C: Context,
+{
     let needs_rotate_key = Some(last_dkg.aggregate_key) != current_aggregate_key;
 
+    // Check if past verification window, if we are, skip verification
+    let dkg_verification_window = context.config().signer.dkg_verification_window;
+    let max_verification_height = last_dkg
+        .started_at_bitcoin_block_height
+        .saturating_add(dkg_verification_window as u64);
+
+    // Get the current bitcoin chain tip to check against the verification window
+    let current_bitcoin_chain_tip = context
+        .get_storage()
+        .get_bitcoin_canonical_chain_tip_ref()
+        .await?
+        .ok_or(Error::NoChainTip)?;
+
+    let past_verification_window = current_bitcoin_chain_tip.block_height > max_verification_height;
+
     let needs_verification = match last_dkg.dkg_shares_status {
-        model::DkgSharesStatus::Unverified => true,
+        model::DkgSharesStatus::Unverified => !past_verification_window,
```

Comment:
Ahh, so this state is possible right now, but unlikely since the block observer marks unverified shares as failed after the verification window. The edge case has to do with things being out of sync sometimes. If we do some of the changes proposed in https://github.com/stacks-sbtc/sbtc/pull/1730, and pass in the bitcoin block here then I think that this state would be impossible.

Is there some other case where we can get to this kind of state with unverified shares past the verification window? cc @matteojug

Reply:
Yeah if we pass the chain tip (and it's the one passed by the block observer) it shouldn't be possible: we could still get failed shares in here because of timing, but since shares cannot go from non-unverified to unverified and we are processing according to (an eventual past) block observer view of things I don't see how we could end up with unverified shares past the verification window. After the changes I would revert this back to `true`.